doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,321 @@
+
+> **_Note_** To achieve the goal of "make it easier to get data in Materialize" we also considered
+building support for Fivetran. We may still do this in the future, but the amount of work to get
+webhooks working is significantly less, and the frequency of data is more suited to the goals of
+Materialize.
+
+## Goals
+
+1. Provide URLs that users can push data to.
+2. Support an extensible syntax similar to `CREATE SOURCE <name> FROM WEBHOOK` that allows
+users to generate the aforementioned URLs.
+3. Handle at least a total of 100 requests per-second across all `WEBHOOK` sources.
+5. Code complete by August 15th, 2023.
+
+
+## Non-Goals
+
+The following are all features we want to support, but are out of scope in the immediate term.
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+* Handling batch requests.
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `WEBHOOK` source, and add a new endpoint to
+`environmentd`'s HTTP server `/api/webhook/:database/:schema/:name`. When a user creates this new
+kind of source, we'll persist the new `Source` in the catalog, use the `storage-client` to create a
+new collection.
+
+When we receive a event we'll use the `mz_adapter::Client` that the HTTP server has to send the raw
+`headers` and `body` to the coordinator, pack the data into a row, and append it to the necessary
+collection using a new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `WEBHOOK`, for which you can
+specify a `BODY FORMAT`, this denotes how we decode the body of a request. To start we'll support
+three formats, `BYTES`, `JSON`, or `TEXT`. By default we will not include the headers of the
+request, they can optionally be added by specifying `INCLUDE HEADERS`.
+
+```
+CREATE SOURCE <name> FROM WEBHOOK
+  BODY FORMAT [BYTES | JSON | TEXT]
+  (INCLUDE HEADERS)?
+;
+```
+
+After executing this request, we'll create a source with the following columns:
+
+| name    | type                        | optional?                                      |
+|---------|-----------------------------|------------------------------------------------|
+| body    | `bytea`, `jsonb`, or `text` | No                                             |
+| headers | `map[text -> text]`         | Yes, present if `INCLUDE HEADERS` is specified |
+
+We'll add a new endpoint to the existing [`base_router`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/environmentd/src/http.rs#L747)
+in `environmentd` with the path: `/api/webhook/:database/:schema/:name`. This follows the existing
```

Comment:
Do URLs support escaping `/` chars if a user ever has a db or schema with a `/` in it? I'm ok with this being a known limitation, but worth understanding before committing.

Reply:
Yes, URLs support escaping `/`!

Reply:
Opened https://github.com/MaterializeInc/materialize/issues/20226 to make sure we can support paths with `/` in them